File: 82-remove-duplicates-from-sorted-list-ii/82-remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *sentinal = new ListNode(INT_MIN, head);
        ListNode *prev = sentinal;
        ListNode *cur = head;
        
        while( cur ) {
            if( cur->next && cur->val == cur->next->val ) {
                ListNode *p = cur->next;
                int dupVal = cur->val;
                delete cur;
                while( p && p->val == dupVal) {
                    ListNode *temp = p;
                    p = p->next;
                    delete temp;
                }    
                prev->next = p;
                cur = p;
            }
            else {
                prev = cur;
                cur = cur->next;
            }
        }
        
        head = sentinal->next;
        delete sentinal;
        return head;
    }
};

File: 82-remove-duplicates-from-sorted-list-ii/82-remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        sentinel = ListNode(float("-inf"), head)
        
        cur = head
        prev = sentinel        
        while cur is not None:
            if cur.next is not None and  cur.val == cur.next.val:
                val = cur.val
                while cur.next is not None and val == cur.next.val:
                    cur = cur.next
                
                prev.next = cur.next
                cur = cur.next
            else:
                prev = cur
                cur = cur.next
        
        return sentinel.next
    
                